content/recipes/authoring/kata-snippets/full-tests.md

```diff
@@ -7,4 +7,59 @@ next: /recipes/authoring/kata-snippets/preloaded/
+- If possible, **reference solution should not be used**. Sometimes there's no other way, but quite often reference solution is simply not necessary. Test cases can be generated with the answer known upfront, and it can save time which would be spent on running the reference solution.
+- **Assertions should be close to the calls of tested solution**, preferrably immediatelly following it. Debugging becomes difficult if invalid answer is verified long after it has been returned.
+- **Random tests should be run after fixed tests.** Not all testing frameworks allow for easy ordering of tests, but fixed tests should run, and eventually fail, before random tests.
+- **Difficulty should be consistent beetwen test runs.** When randomized ranges are very large, then it's possible that one user receives many small, easy inputs and their solution passes, while another user receives many large, more difficult inputs, and their solution fails or times out. Make sure that your random tests are built in a way which elimnates such situation. If you want to test difficult inputs, split test cases into a set of easy ones and a set of diffficult ones, and tets them separately.
+- **Make debugging of random tests easy.** Sometimes, when random tests fail, they are very difficult to debug because input is very large and cannot be easily printed. If possible, Try to compose your test suite in such a way that when test fail, they fail on some small, easy, debuggable input, before they fail on a large one, which is difficult to print. 
+
+## Performance Tests
+
+Some kata require solutions to be fast, or to fall into some performance category. For example, author might want to accept only solutions of complexity `O(n)`, and reject `O(n^2)` ones. Building such test suite is not always easy!
+
+- **Performance tests can be implemented in terms of random tests.** Usually, it's not reqired to have separate sections for random tests and performance tests. Testing of performance can be driven by size of random inputs. But still, to make debugging easier, it can be worth in some cases to have a set of separate random tests with small inputs.
+- If difficulty of a tested kata is related to the size of the input, it's usually **better to have a few tests with large input, that thousands of tests with small input.** 200 tests with very large numbers or arrays is usually better than 10_000 tests with small ones (but see remarks on size of inputs above).
+- **Difference beetwen accepted and rejected solutions should be easy to spot.** Ideally, tests for accepted solutions should complete very fast, and tests for rejected solutions should take very long. Otherwise, you risk that solutions with valid complexity characteristics will time out, and users will be frustrated looking for micro-optimizations. This approach usually calls for very large inputs, so be careful! 
```

Comment:
```suggestion
- **The difference between accepted and rejected solutions should be easy to spot.** Ideally, accepted solutions should complete well under the time limit, while rejected solutions should time out consistently. Otherwise, you risk that solutions with valid complexity characteristics will time out, and users will be frustrated looking for micro-optimizations. This approach usually calls for very large inputs, so be careful!
```